pkg/ingressCache/ingressCache.go

```diff
@@ -0,0 +1,124 @@
+/*
+Copyright 2019 Iguazio Systems Ltd.
+
+Licensed under the Apache License, Version 2.0 (the "License") with
+an addition restriction as set forth herein. You may not use this
+file except in compliance with the License. You may obtain a copy of
+the License at http://www.apache.org/licenses/LICENSE-2.0.
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+implied. See the License for the specific language governing
+permissions and limitations under the License.
+
+In addition, you may not use the software for any purposes that are
+illegal under applicable law, and the grant of the foregoing license
+under the Apache 2.0 license is conditioned upon your compliance with
+such restriction.
+*/
+
+package ingresscache
+
+import (
+	"sync"
+
+	"github.com/nuclio/errors"
+	"github.com/nuclio/logger"
+)
+
+type IngressHostsTree interface {
+	SetFunctionName(path string, function string) error // will overwrite existing values if exists
+	DeleteFunctionName(path string, function string) error
+	GetFunctionName(path string) ([]string, error)
+	IsEmpty() bool
+}
+
+type IngressHostCache interface {
+	Set(host string, path string, function string) error // will overwrite existing values if exists
+	Delete(host string, path string, function string) error
+	Get(host string, path string) ([]string, error)
+}
+
+type IngressCache struct {
+	syncMap *sync.Map
+	logger  logger.Logger
+}
+
+func NewIngressCache(logger logger.Logger) *IngressCache {
+	return &IngressCache{
+		syncMap: &sync.Map{},
+		logger:  logger,
+	}
+}
+
+func (ic *IngressCache) Set(host, path, function string) error {
+	urlTree, exists := ic.syncMap.Load(host)
+	if !exists {
+		urlTree = NewSafeTrie()
+		ic.syncMap.Store(host, urlTree)
```

Comment:
Why not store the tree in the syncMap at the end? then you don't need to delete it on every error